src/file_pipe_log/reader.rs

```diff
@@ -64,17 +65,59 @@ impl<F: FileSystem> LogItemBatchFileReader<F> {
     /// Returns the next [`LogItemBatch`] in current opened file. Returns
     /// `None` if there is no more data or no opened file.
     pub fn next(&mut self) -> Result<Option<LogItemBatch>> {
-        if self.valid_offset < self.size {
+        // @lucasliang.
+        // TODO: [Fulfilled in writing progress when DIO is open.]
+        // We should also consider that there might exists broken blocks when DIO
+        // is open, and the following reading strategy should tolerate reading broken
+        // blocks until it finds an accessible header of `LogBatch`.
+        while self.valid_offset < self.size {
+            debug_assert!(self.file_context.is_some());
+            let data_layout = self.file_context.as_ref().unwrap().format.data_layout();
             if self.valid_offset < LOG_BATCH_HEADER_LEN {
                 return Err(Error::Corruption(
                     "attempt to read file with broken header".to_owned(),
                 ));
             }
-            let (footer_offset, compression_type, len) = LogBatch::decode_header(&mut self.peek(
-                self.valid_offset,
-                LOG_BATCH_HEADER_LEN,
-                0,
-            )?)?;
+            let (footer_offset, compression_type, len) = {
+                match data_layout {
+                    DataLayout::NoAlignment | DataLayout::Alignment(0) => {
+                        let (offset, t, l) = LogBatch::decode_header(&mut self.peek(
+                            self.valid_offset,
+                            LOG_BATCH_HEADER_LEN,
+                            0,
+                        )?)?;
+                        (offset, t, l)
+                    }
+                    DataLayout::Alignment(fs_block_size) => {
+                        // In DataLayout::Alignment mode, tail data in this block maybe
+                        // contain an valid header of the adjacent LogBatch or paddings
+                        // with `0`.
+                        match LogBatch::decode_header(&mut self.peek(
+                            self.valid_offset,
+                            round_up(self.valid_offset, fs_block_size as usize) - self.valid_offset
```

Comment:
Why reading it differently? You should always read `LOG_BATCH_HEADER_LEN`, but the `hint` might be different.

A better way to write it would be:
```rust
fn padding_size() -> usize; // returns the number of bytes needed to pad the offset to given alignment.
let padding = padding_size(data_layout.alignment(), self.valid_offset);
let (..) = match LogBatch::decode_header(self.valid_offset, LOG_BATCH_HEADER_LEN, padding) {
  Err(e) if padding >= LOG_BATCH_HEADER_LEN {
    // check padding is zeros.
   continue;
  }
  Ok(r) => r,
  e @ Err(_) => return r,
}
```